File: server/ubudesk_server/protocol.py
"""UbuDesk wire protocol v1: framing and message helpers.

Framing (see docs/PROTOCOL.md):

    | type: u8 | length: u32 big-endian | payload: `length` bytes |

Types:
    0x01 CONTROL  - UTF-8 JSON object with a "t" field.
    0x02 VIDEO    - | pts_us: u64 BE | flags: u8 | H.264 Annex-B access unit |
                    flags bit0 = keyframe (IDR), bit1 = contains SPS/PPS.

Max payload is 8 MiB; a peer announcing more must be disconnected.
"""

from __future__ import annotations

import asyncio
import json
import struct
from typing import Any

MAX_PAYLOAD = 8 * 1024 * 1024

TYPE_CONTROL = 0x01
TYPE_VIDEO = 0x02

VIDEO_FLAG_KEYFRAME = 0x01
VIDEO_FLAG_CONFIG = 0x02

_HEADER = struct.Struct(">BI")
_VIDEO_HEADER = struct.Struct(">QB")


class ProtocolError(Exception):
    """Fatal protocol violation; the connection must be closed."""


def encode_control(msg: dict[str, Any]) -> bytes:
    payload = json.dumps(msg, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
    if len(payload) > MAX_PAYLOAD:
        raise ProtocolError("control message too large")
    return _HEADER.pack(TYPE_CONTROL, len(payload)) + payload


def encode_video(pts_us: int, flags: int, data: bytes) -> bytes:
    payload_len = _VIDEO_HEADER.size + len(data)
    if payload_len > MAX_PAYLOAD:
        raise ProtocolError("video frame too large")
    return _HEADER.pack(TYPE_VIDEO, payload_len) + _VIDEO_HEADER.pack(pts_us, flags) + data


def decode_control(payload: bytes) -> dict[str, Any]:
    try:
        msg = json.loads(payload.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise ProtocolError(f"bad control payload: {exc}") from exc
    if not isinstance(msg, dict) or not isinstance(msg.get("t"), str):
        raise ProtocolError("control message must be a JSON object with a string 't'")
    return msg


def decode_video(payload: bytes) -> tuple[int, int, bytes]:
    """Return (pts_us, flags, annexb_data)."""
    if len(payload) < _VIDEO_HEADER.size:
        raise ProtocolError("video payload too short")
    pts_us, flags = _VIDEO_HEADER.unpack_from(payload)
    return pts_us, flags, payload[_VIDEO_HEADER.size :]


def parse_header(header: bytes) -> tuple[int, int]:
    """Parse the 5-byte frame header; returns (type, length)."""
    if len(header) != _HEADER.size:
        raise ProtocolError("truncated header")
    ftype, length = _HEADER.unpack(header)
    if length > MAX_PAYLOAD:
        raise ProtocolError(f"payload length {length} exceeds 8 MiB limit")
    return ftype, length


HEADER_SIZE = _HEADER.size  # 5


async def read_frame(reader: asyncio.StreamReader) -> tuple[int, bytes]:
    """Read one frame; returns (type, payload). Raises ProtocolError / IncompleteReadError."""
    header = await reader.readexactly(_HEADER.size)
    ftype, length = parse_header(header)
    payload = await reader.readexactly(length) if length else b""
    return ftype, payload


# ---------------------------------------------------------------------------
# Annex-B helpers (shared by tests and the test source)


def annexb_nal_types(data: bytes) -> list[int]:
    """Return the NAL unit types found in an Annex-B byte stream, in order."""
    types: list[int] = []
    i = 0
    n = len(data)
    while i + 3 < n:
        if data[i] == 0 and data[i + 1] == 0:
            if data[i + 2] == 1:
                types.append(data[i + 3] & 0x1F)
                i += 4
                continue
            if data[i + 2] == 0 and i + 4 < n and data[i + 3] == 1:
                types.append(data[i + 4] & 0x1F)
                i += 5
                continue
        i += 1
    return types


def contains_sps_pps(data: bytes) -> bool:
    ts = annexb_nal_types(data)
    return 7 in ts and 8 in ts


def contains_idr(data: bytes) -> bool:
    return 5 in annexb_nal_types(data)

File: android/app/src/main/java/app/ubudesk/client/net/Protocol.kt
package app.ubudesk.client.net

import org.json.JSONObject
import java.io.DataInputStream
import java.io.EOFException
import java.io.InputStream
import java.io.OutputStream
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

/**
 * UbuDesk wire protocol v1 (see docs/PROTOCOL.md).
 *
 * Framing: | type: u8 | length: u32 BE | payload |
 *  - 0x01 CONTROL: UTF-8 JSON with a "t" field.
 *  - 0x02 VIDEO:   | pts_us: u64 BE | flags: u8 | H.264 Annex-B access unit |
 */
object Protocol {
    const val VERSION = 1
    const val TYPE_CONTROL = 0x01
    const val TYPE_VIDEO = 0x02
    const val MAX_PAYLOAD = 8 * 1024 * 1024
    const val VIDEO_HEADER_SIZE = 9

    const val FLAG_KEYFRAME = 0x01
    const val FLAG_CONFIG = 0x02

    class ProtocolException(message: String) : Exception(message)

    sealed interface Frame
    data class Control(val msg: JSONObject) : Frame {
        val t: String get() = msg.optString("t")
    }
    data class Video(val ptsUs: Long, val flags: Int, val data: ByteArray) : Frame {
        val isKeyframe: Boolean get() = flags and FLAG_KEYFRAME != 0
        val hasConfig: Boolean get() = flags and FLAG_CONFIG != 0

        override fun equals(other: Any?): Boolean =
            other is Video && ptsUs == other.ptsUs && flags == other.flags &&
                data.contentEquals(other.data)

        override fun hashCode(): Int = 31 * ptsUs.hashCode() + flags
    }

    fun encodeControl(msg: JSONObject): ByteArray {
        val payload = msg.toString().toByteArray(StandardCharsets.UTF_8)
        require(payload.size <= MAX_PAYLOAD) { "control message too large" }
        val buf = ByteBuffer.allocate(5 + payload.size)
        buf.put(TYPE_CONTROL.toByte())
        buf.putInt(payload.size)
        buf.put(payload)
        return buf.array()
    }

    /** Parses a complete frame from [wire]; used by tests and readFrame. */
    fun decodeFrame(wire: ByteArray): Frame {
        if (wire.size < 5) throw ProtocolException("truncated header")
        val buf = ByteBuffer.wrap(wire)
        val type = buf.get().toInt() and 0xFF
        val length = buf.int
        if (length > MAX_PAYLOAD) throw ProtocolException("payload exceeds 8 MiB")
        if (wire.size - 5 < length) throw ProtocolException("truncated payload")
        val payload = ByteArray(length)
        buf.get(payload)
        return decodePayload(type, payload)
    }

    fun decodePayload(type: Int, payload: ByteArray): Frame = when (type) {
        TYPE_CONTROL -> {
            val text = String(payload, StandardCharsets.UTF_8)
            val json = try {
                JSONObject(text)
            } catch (e: Exception) {
                throw ProtocolException("bad control JSON: ${e.message}")
            }
            if (!json.has("t")) throw ProtocolException("control message missing 't'")
            Control(json)
        }
        TYPE_VIDEO -> {
            if (payload.size < VIDEO_HEADER_SIZE) throw ProtocolException("video payload too short")
            val buf = ByteBuffer.wrap(payload)
            val pts = buf.long
            val flags = buf.get().toInt() and 0xFF
            val data = ByteArray(payload.size - VIDEO_HEADER_SIZE)
            buf.get(data)
            Video(pts, flags, data)
        }
        else -> throw ProtocolException("unknown frame type $type")
    }

    /** Blocking read of one frame from a stream. Returns null on unknown types (skipped). */
    fun readFrame(input: DataInputStream): Frame? {
        val type = try {
            input.readUnsignedByte()
        } catch (e: EOFException) {
            throw e
        }
        val length = input.readInt()
        if (length < 0 || length > MAX_PAYLOAD) throw ProtocolException("bad frame length $length")
        val payload = ByteArray(length)
        input.readFully(payload)
        return when (type) {
            TYPE_CONTROL, TYPE_VIDEO -> decodePayload(type, payload)
            else -> null // unknown type: ignore per spec
        }
    }

    fun writeControl(output: OutputStream, msg: JSONObject) {
        output.write(encodeControl(msg))
        output.flush()
    }

    // ---- message builders --------------------------------------------------

    fun hello(clientId: String, name: String, appVersion: String, w: Int, h: Int, dpi: Int, refresh: Int): JSONObject =
        JSONObject()
            .put("t", "hello")
            .put("proto", VERSION)
            .put("client_id", clientId)
            .put("name", name)
            .put("app_version", appVersion)
            .put(
                "screen",
                JSONObject().put("w", w).put("h", h).put("dpi", dpi).put("refresh", refresh),
            )
            .put("codecs", org.json.JSONArray().put("h264"))

    fun authPin(pin: String): JSONObject =
        JSONObject().put("t", "auth").put("method", "pin").put("pin", pin)

    fun authToken(token: String): JSONObject =
        JSONObject().put("t", "auth").put("method", "token").put("token", token)

    fun start(width: Int, height: Int, fps: Int, bitrateKbps: Int, mode: String, touchMode: String): JSONObject =
        JSONObject()
            .put("t", "start")
            .put("width", width and 1.inv())
            .put("height", height and 1.inv())
            .put("fps", fps)
            .put("bitrate_kbps", bitrateKbps)
            .put("mode", mode)
            .put("touch_mode", touchMode)

    fun touch(action: String, id: Int, x: Double, y: Double): JSONObject =
        JSONObject().put("t", "touch").put("a", action).put("id", id).put("x", x).put("y", y)

    fun mouse(action: String, button: String?, x: Double, y: Double): JSONObject {
        val o = JSONObject().put("t", "mouse").put("a", action).put("x", x).put("y", y)
        if (button != null) o.put("b", button)
        return o
    }

    fun scroll(dx: Double, dy: Double): JSONObject =
        JSONObject().put("t", "scroll").put("dx", dx).put("dy", dy)

    fun key(code: Int, down: Boolean): JSONObject =
        JSONObject().put("t", "key").put("code", code).put("down", down)

    fun text(s: String): JSONObject = JSONObject().put("t", "text").put("s", s)

    fun ping(seq: Int, ts: Long): JSONObject =
        JSONObject().put("t", "ping").put("seq", seq).put("ts", ts)

    fun idr(): JSONObject = JSONObject().put("t", "idr")

    fun stats(fps: Double, decodeMs: Double, dropped: Int): JSONObject =
        JSONObject().put("t", "stats").put("fps", fps).put("decode_ms", decodeMs).put("dropped", dropped)

    fun bye(): JSONObject = JSONObject().put("t", "bye")
}
